sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -844,101 +843,50 @@ public final synchronized BigtableSource splitAtFraction(double fraction) {
 
-    public Sink(String tableId, BigtableService bigtableService) {
+    public BigtableWriterFn(String tableId, BigtableService bigtableService) {
       this.tableId = checkNotNull(tableId, "tableId");
       this.bigtableService = checkNotNull(bigtableService, "bigtableService");
-    }
-
-    public String getTableId() {
-      return tableId;
-    }
-
-    public BigtableService getBigtableService() {
-      return bigtableService;
-    }
-
-    @Override
-    public String toString() {
-      return MoreObjects.toStringHelper(Sink.class)
-          .add("bigtableService", bigtableService)
-          .add("tableId", tableId)
-          .toString();
-    }
-
-    ///////////////////////////////////////////////////////////////////////////////
-    private final String tableId;
-    private final BigtableService bigtableService;
-
-    @Override
-    public WriteOperation<KV<ByteString, Iterable<Mutation>>, Long> createWriteOperation(
-        PipelineOptions options) {
-      return new BigtableWriteOperation(this);
-    }
-
-    /** Does nothing, as it is redundant with {@link Write#validate}. */
-    @Override
-    public void validate(PipelineOptions options) {}
-  }
-
-  private static class BigtableWriteOperation
-      extends WriteOperation<KV<ByteString, Iterable<Mutation>>, Long> {
-    private final Sink sink;
-
-    public BigtableWriteOperation(Sink sink) {
-      this.sink = sink;
+      this.failures = new ConcurrentLinkedQueue<>();
     }
 
     @Override
-    public Writer<KV<ByteString, Iterable<Mutation>>, Long> createWriter(PipelineOptions options)
-        throws Exception {
-      return new BigtableWriter(this);
+    public void startBundle(Context c) throws Exception {
+      bigtableWriter = bigtableService.openForWriting(tableId);
+      recordsWritten = 0;
     }
 
     @Override
-    public void initialize(PipelineOptions options) {}
-
-    @Override
-    public void finalize(Iterable<Long> writerResults, PipelineOptions options) {
-      long count = 0;
-      for (Long value : writerResults) {
-        value += count;
-      }
-      logger.debug("Wrote {} elements to BigtableIO.Sink {}", sink);
+    public void processElement(ProcessContext c) throws Exception {
+      checkForFailures();
+      Futures.addCallback(
+          bigtableWriter.writeRecord(c.element()), new WriteExceptionCallback(c.element()));
+      ++recordsWritten;
     }
 
     @Override
-    public Sink getSink() {
-      return sink;
+    public void finishBundle(Context c) throws Exception {
+      bigtableWriter.close();
+      bigtableWriter = null;
+      checkForFailures();
+      logger.info("Wrote {} records", recordsWritten);
     }
 
     @Override
-    public Coder<Long> getWriterResultCoder() {
-      return VarLongCoder.of();
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.add(DisplayData.item("tableId", tableId)
+        .withLabel("Table ID"));
```

Comment:
And bigtableOptions: https://github.com/ianzhou1/incubator-beam/blob/10f8e5d8fc9ebe0c608bfd935e2b6c0131352df8/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java#L270